cloudinit/distros/__init__.py

```
@@ -139,9 +157,88 @@ def _unpickle(self, ci_pkl_version: int) -> None:
+                            LOG.error(
+                                "Cannot install packages under %s as it is "
+                                "not a supported package manager!",
+                                package_manager,
+                            )
+            elif isinstance(entry, str):
+                generic_packages.add(entry)
+            else:
+                raise ValueError(
+                    "Invalid 'packages' yaml specification. "
+                    "Check schema definition."
+                )
+        return dict(packages_by_manager), generic_packages
+
     def install_packages(self, pkglist):
-        raise NotImplementedError()
+        error_message = (
+            "Failed to install the following packages: %s. "
+            "See associated package manager logs for more details."
+        )
+        # If an entry hasn't been included with an explicit package name,
+        # add it to a 'generic' list of packages
+        (
+            packages_by_manager,
+            generic_packages,
+        ) = self._extract_package_by_manager(pkglist)
+
+        # First install packages using package manager(s)
+        # supported by the distro
+        uninstalled = []
+        for manager in self.package_managers:
+            to_try = (
+                packages_by_manager.get(manager.__class__, set())
+                | generic_packages
+            )
+            uninstalled = manager.install_packages(to_try)
+            failed = {
+                pkg for pkg in uninstalled if pkg not in generic_packages
+            }
+            if failed:
+                LOG.error(error_message, failed)
+            generic_packages = set(uninstalled)
+
+        # Now attempt any specified package managers not explicitly supported
+        # by distro
+        for manager, packages in packages_by_manager.items():
+            if manager.name not in known_package_managers:
+                LOG.error(
```

It doesn't look like we have unittest/integration coverage for this error, either would be fine.

I think we actually want to raise an error or continue here otherwise we continue after this and fail below with an ugly traceback on 234 because presumably `manager` is not known/supported in this image.